modules/sitemaps/sitemaps-provider-decorator.php

```diff
@@ -59,29 +79,113 @@ public function get_url_list( $page_num, $object_subtype = '' ) {
 	/**
 	 * Gets the max number of pages available for the object type.
 	 *
-	 * Returns 0 for all languages but the default language for untranslated post types
-	 * and untranslated taxonomies to avoid to duplicate the corresponding sitemaps.
-	 *
 	 * @since 2.8
 	 *
 	 * @param string $object_subtype Optional. Object subtype. Default empty.
 	 * @return int Total number of pages.
 	 */
 	public function get_max_num_pages( $object_subtype = '' ) {
-		switch ( $this->provider->name ) {
-			case 'posts':
-				if ( ! PLL()->model->is_translated_post_type( $object_subtype ) && PLL()->model->options['default_lang'] !== $this->lang ) {
-					return 0;
-				}
-				break;
-			case 'taxonomies':
-				if ( ! PLL()->model->is_translated_taxonomy( $object_subtype ) && PLL()->model->options['default_lang'] !== $this->lang ) {
-					return 0;
-				}
-				break;
+		return $this->provider->get_max_num_pages( $object_subtype );
+	}
+
+	/**
+	 * Get active languages for the sitemap.
+	 *
+	 * @since 2.8
+	 */
+	protected function get_active_languages() {
+		$languages = $this->model->get_languages_list();
+		if ( wp_list_filter( $languages, array( 'active' => false ) ) ) {
+			return wp_list_pluck( wp_list_filter( $languages, array( 'active' => false ), 'NOT' ), 'slug' );
 		}
+		return wp_list_pluck( $languages, 'slug' );
+	}
 
-		return $this->provider->get_max_num_pages( $object_subtype );
+	/**
+	 * Filters the query arguments to add the language.
+	 *
+	 * @since 2.8
+	 *
+	 * @param array $args Sitemap provider WP_Query or WP_Term_Query arguments.
+	 * @return array
+	 */
+	public function query_args( $args ) {
+		if ( isset( $this->filter_lang ) ) {
+			$args['lang'] = $this->filter_lang;
+		}
+		return $args;
+	}
+
+	/**
+	 * Gets data for a given sitemap type.
+	 *
+	 * @since 2.8
+	 *
+	 * @param string $object_subtype_name Object subtype name if any.
+	 * @param string $lang                Optionnal language name.
+	 * @return array
+	 */
+	protected function get_sitemap_data( $object_subtype_name, $lang = '' ) {
+		$object_subtype_name = (string) $object_subtype_name;
+
+		if ( ! empty( $lang ) ) {
+			$this->filter_lang = $lang;
+		}
+
+		$return = array(
+			'name'  => implode( '-', array_filter( array( $object_subtype_name, $lang ) ) ),
+			'pages' => $this->get_max_num_pages( $object_subtype_name ),
+		);
+
+		unset( $this->filter_lang );
+		return $return;
+	}
+
+	/**
+	 * Gets data about each sitemap type.
+	 *
+	 * @since 2.8
+	 *
+	 * @return array[] Array of sitemap types including object subtype name and number of pages.
+	 */
+	public function get_sitemap_type_data() {
+		$sitemap_data = array();
+
+		add_filter( 'wp_sitemaps_posts_query_args', array( $this, 'query_args' ) );
+		add_filter( 'wp_sitemaps_taxonomies_query_args', array( $this, 'query_args' ) );
```

Comment:
These hooks are registered once per sitemap provider. That means that the second provider will also activate the 1st provider's hooks, and the third will activate the 1st and 2nd providers' hooks as well. I've noticed that as the current state of the code you've taken care that these former providers' hooks will have no effect, but I'm worried this implicit behavior could be overlooked in the future.

Also, these hooks are both called from two functions, and I'm not sure you want to call them in both cases, or just on one for each. See: 
- https://github.com/GoogleChromeLabs/wp-sitemaps/blob/c8c36818811fd7e2d339b86ddded08f532a075e4/inc/providers/class-wp-sitemaps-posts.php#L90-L93
- https://github.com/GoogleChromeLabs/wp-sitemaps/blob/c8c36818811fd7e2d339b86ddded08f532a075e4/inc/providers/class-wp-sitemaps-posts.php#L176-L180

An alternative to using hooks here, would be to decorate the `WP_Sitemaps_Post::get_posts_query_args` and `WP_Sitemaps_Taxonomies::get_taxonomies_query_args` methods and filter the lists here. Next step could be to ask the WP Sitemaps team to make merge these two functions in one (`WP_Sitemaps_Provider::get_query_args` for example, so we could only manage this in one place (and the other plugins developers could use this method name to be compatible with our decorator).